content/blog/2020/07/2020-07-27-machine-learning-plugin-coding-phase2.adoc

```diff
@@ -0,0 +1,160 @@
+Machine learning plugin has successfully passed the end to end test. A feature for a code editor is needed for further discussion/analysis as we have done a simple editor that may be useful in other ways in the future. link:https://github.com/jenkinsci/machine-learning-plugin/pull/35[PR#35]
+
+==== Main features of Machine Learning plugin
+
+- Run Jupyter notebook, (Zeppelin) JSON and Python files
+- Run Python code directly
+- Convert Jupyter Notebooks to Python and JSON
+- Configure IPython kernel properties
+- Support to execute Notebooks/Python on Agent
+- Support for Windows and Linux
+
+====== Upcoming features
+
+- Extract graph/map/images from the code
+- Save artifacts according to the step name
+- Generate reports for corresponding build
+
+====== Future improvements
+
+- Usage of JupyterRestClient
+- Support for multiple language kernels
+
+==== Docker agent
+
+The following dockerfile can be used to build the docker container as an agent for machine learning plugin. This docker agent can be used to run  notebooks or python scripts.
+
+.Dockerfile
+
+```
+FROM jenkins/agent:latest
+
+MAINTAINER Loghi <loghijiaha@gmail.com>
+
+USER root
+
+RUN apt update && apt install --no-install-recommends python3 -y \
+    python3-pip \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY requirements.txt /requirements.txt
+
+RUN pip3 install --upgrade pip setuptools && \
+    pip3 install --no-cache-dir -r /requirements.txt && \
+    ln -sf /usr/bin/python3 /usr/bin/python && \
+    ln -sf /usr/bin/pip3 /usr/bin/pip
+
+USER jenkins
+```
+
+
+==== Ported to Python 3
+
+As discussed in the previous meeting, we concluded that the plugin should support Python 3 as Python 2.7+ has been deprecated since the beginning of 2020. Pull request for docker agent should be also ported to Python 3 support.
+
+==== Jupyter Rest Client API
+
+The Jupyter Notebook server API seemed to be promising that it can be also used to run notebooks and codes. There were 3 api implementations that were merged in the master. But we had to focus on what was proposed in the design document and had to finish all must-have issues/works. Jupyter REST client was left for future implementation. It is also a good start to contribute to the plugin from the community.
+
+
+==== Fixed bugs for running in agent
+
+There were a few bugs related to the file path of notebooks while building a job. The most problem was caused by the python dependencies needed to connect to a IPython kernel. All issues/bugs were fixed before the timeline given.
```

Comment:
Instead of "The most problem" I think "The main problem", or "The major problem" perhaps?

Reply:
(And well done on keeping ahead of the schedule @Loghijiaha !)